cpp/example_code/s3/CMakeLists.txt

```diff
@@ -1,49 +1,104 @@
-list(APPEND EXAMPLES "list_buckets")
-list(APPEND EXAMPLES "list_buckets_disabling_dns_cache")
-list(APPEND EXAMPLES "list_objects")
-list(APPEND EXAMPLES "list_objects_with_aws_global_region")
-list(APPEND EXAMPLES "put_bucket_policy")
-list(APPEND EXAMPLES "put_object")
-list(APPEND EXAMPLES "put_object_async")
-list(APPEND EXAMPLES "put_object_buffer")
-list(APPEND EXAMPLES "put_website_config")
-list(APPEND EXAMPLES "s3-demo")
-
-# The executables to build
-foreach(EXAMPLE IN LISTS EXAMPLES)
-  add_executable(${EXAMPLE} ${EXAMPLE}.cpp)
-  target_link_libraries(${EXAMPLE} ${AWSSDK_LINK_LIBRARIES} ${AWSSDK_PLATFORM_DEPS})
+# Add the code example-specific header files.
+file(GLOB AWSDOC_S3_HEADERS
+	"include/awsdoc/s3/*.h"
+)
+
+# Add the code example-specific source files.
+file(GLOB AWSDOC_S3_SOURCE
+	"*.cpp"
+)
+
+# Check whether the target system is Windows, including Win64.
+if(WIN32)
+	# Check whether the compiler is some version of Microsoft Visual C++, or another compiler simulating C++.
+	if(MSVC)
+		source_group("Header Files\\awsdoc\\s3" FILES ${AWSDOC_S3_HEADERS})
+		source_group("Source Files" FILES ${AWSDOC_S3_SOURCE})
+	endif(MSVC)
+endif()
+
+foreach(file ${AWSDOC_S3_SOURCE})
+	get_filename_component(EXAMPLE ${file} NAME_WLE)
+
+	# Build the code example executables.
+	set(EXAMPLE_EXE run_${EXAMPLE})
+
+	add_executable(${EXAMPLE_EXE} ${AWSDOC_S3_HEADERS} ${file})
+
+	if(MSVC AND BUILD_SHARED_LIBS)
+		target_compile_definitions(${EXAMPLE_EXE} PUBLIC "USE_IMPORT_EXPORT")
+		target_compile_definitions(${EXAMPLE_EXE} PRIVATE "AWSDOC_S3_EXPORTS")
+	endif()
+
+	target_include_directories(${EXAMPLE_EXE} PUBLIC 
+		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+		$<INSTALL_INTERFACE:include>)
+	target_link_libraries(${EXAMPLE_EXE} ${AWSSDK_LINK_LIBRARIES} 
+		${AWSSDK_PLATFORM_DEPS})
+
+	if(BUILD_TESTING)
+	    # Enable testing for this directory and below.
+		enable_testing()
+
+		# Build the code example libraries.
+		set(EXAMPLE_LIB ${EXAMPLE})
+
+		add_library(${EXAMPLE_LIB} ${AWSDOC_S3_HEADERS} ${file})
+
+		if(MSVC AND BUILD_SHARED_LIBS)
+            target_compile_definitions(${EXAMPLE_LIB} PUBLIC "USE_IMPORT_EXPORT")
+            target_compile_definitions(${EXAMPLE_LIB} PRIVATE "AWSDOC_S3_EXPORTS")
+        endif()
+
+		target_include_directories(${EXAMPLE_LIB} PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+            $<INSTALL_INTERFACE:include>)
+        target_link_libraries(${EXAMPLE_LIB} ${AWSSDK_LINK_LIBRARIES} 
+			${AWSSDK_PLATFORM_DEPS})
+
+		# Build the code example unit tests.
+        set(EXAMPLE_TEST test_${EXAMPLE})
+
+		add_executable(${EXAMPLE_TEST} tests/test_${EXAMPLE}.cpp)
+
+        target_include_directories(${EXAMPLE_TEST} PUBLIC
+            $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+            $<INSTALL_INTERFACE:include>)
+        target_link_libraries(${EXAMPLE_TEST} ${EXAMPLE_LIB})
+        add_test(${EXAMPLE_TEST} ${EXAMPLE_TEST})

```

Comment:
I'd change these lines to:
```cmake
set(EXAMPLE_TEST_FILE ${CMAKE_CURRENT_SOURCE_DIR}/tests/test_${EXAMPLE}.cpp)
if(EXISTS ${EXAMPLE_TEST_FILE})
	add_executable(${EXAMPLE_TEST} ${EXAMPLE_TEST_FILE})

	target_include_directories(${EXAMPLE_TEST} PUBLIC
		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
		$<INSTALL_INTERFACE:include>)
	target_link_libraries(${EXAMPLE_TEST} ${EXAMPLE_LIB})
	add_test(${EXAMPLE_TEST} ${EXAMPLE_TEST})
endif()
```
1. First, try to find the testing file, don't create testing executables if it's not exists.
2. Ideally, don't build the library for `create_bucket` if testing file do not exist. It's not necessary to have `add_library(${EXAMPLE_LIB} ${AWSDOC_S3_HEADERS} ${file})` if there is no test for that.